devlib/utils/ssh.py

```diff
@@ -299,6 +313,498 @@ def __init__(self,
         self.conn = ssh_get_shell(host, username, password, None, port, timeout, True, original_prompt)
 
 
+class Gem5Connection(TelnetConnection):
+
+    def __init__(self,
+                 platform,
+                 host=None,
+                 username=None,
+                 password=None,
+                 port=None,
+                 timeout=None,
+                 password_prompt=None,
+                 original_prompt=None,
+                 ):
+        # TODO: need to insert check that host is the actual place
+        # where devlib is now running
+        self.host = host
+        if username is not None and username != 'root':
+            raise ValueError('User should be root in gem5!')
+        if password is not None and password != '':
+            raise ValueError('No password needed in gem5!')
+        self.username = 'root'
+        self.is_rooted = True
+        self.password = None
+        self.port = -1
+        # Long timeouts to account for gem5 being slow
+        # Can be overriden if the given timeout is longer
+        self.default_timeout = 3600
+        if timeout is not None:
+            if timeout > self.default_timeout:
+                logger.info('Overwriting the default timeout of gem5 ({})'
+                                 ' to {}'.format(self.default_timeout, timeout))
+                self.default_timeout = timeout
+            else :
+                logger.info('Ignoring the given timeout --> gem5 needs longer timeouts')
+        self.ready_timeout = self.default_timeout * 3
+        # Intermediate directory to push things to gem5 using VirtIO
+        self.gem5_interact_dir = None
+        # Counterpart in gem5
+        self.gem5_input_dir = '/mnt/host/'
+        # Directory to store things from gem5 on the host
+        self.gem5_out_dir = None
+        self.busybox_wrapper = ''
+        self.m5_path = None
```

Comment:
But then the init method here should not contain the platform either?
So would something like the following be okay:

- if you pass a platform during initialisation of the connection, the connection connects to it automatically (_connect_gem5) and those attributes are set by that method
- or you pass no platform and the values for these attributes (eg gem5_out_dir) are provided by the init arguments and another connection method is added, that allows you to connect to a gem5 simulation (not a platform)?

And then I'll add checks to check if an attribute is actually set before you use it?